### DISCUSSION TITLE
Async JSON read/write/(de)serialize

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/runtime#66885] [Q&A]
Does it make sense to make JSON operations async, with `ValueTask` and writing to a `PipeWriter`? I'm concerned about potential performance hit of all the allocations of async state machines. Has anybody looked into it at all?

**Comment by FiniteReality** [comment]
I believe that if a method returns ValueTask, you can expect it to run completely synchronously without allocating a state machine for as long as it can. In this case, it would be specifically when the PipeWriter buffer is empty.

**Reply by TahirAhmadov** [reply]
Did you mean "completely synchronously"? I think a `Task` is not allocated, but any time you have `async` method, it allocates a state machine, even if it returns a `ValueTask`. In some cases, I can just return the `ValueTask` without using `async` and that would be zero allocations, but it's not possible everywhere.

**Reply by FiniteReality** [reply]
I did, my apologies. Looking at the code, there seems to only be the allocation of a single Task<T> within the state machine, and the implementation avoids further allocations.

**Reply by TahirAhmadov** [reply]
Yes, but the state machine itself is a heap object which is always allocated. Meaning, with `Task`, it's actually two allocations per `async` method call.

**Reply by FiniteReality** [reply]
It doesn't seem to be a heap object: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4CsA3IYRgMzlakDCpA3oaT+UxgA5SANQCGAGwCuMDElIBZABQBKTt14aKATkUAiUbuW0CGzVtIA5FcdO9te4IZu3y5q0fW37usE8+kAX39/RhEJaVlLFTUTFwwAdlIAExgAM1FJcQAXZx4gggCgA

**Reply by TahirAhmadov** [reply]
Oh yes, you are right - I was looking at IL generated by `Task` assumed it did a heap alloc with `ValueTask`, as well.
I guess the remaining question is, does the non-allocation overhead add up to making it more expensive than the benefits of writing JSON async.

**Reply by TahirAhmadov** [reply]
Clarification: that was old, .NET 4.8 code which allocates a heap object for state machine. In .NET 6, state machines are always a struct, regardless of `Task` or `ValueTask`.

**Comment by TahirAhmadov** [comment]
I ran a quick benchmark and it looks like when writing to a `StringBuilder`, everything else being equal, (.NET 6) a collection of 10,000 DTOs is JSON-written in: (average of 10 runs)
Sync:  00:00:00.0309208
Async: 00:00:00.0651994

**Comment by TahirAhmadov** [comment]
To answer my own question: it does not seem to be worth it to make methods of a JsonWriter-like class asynchonous, even with ValueTask, and this includes HtmlTextWriter. I've actually written an async HtmlTextWriter implementation and code to use it, and am going to revert to synchronous. The performance penalty, rendering a moderately sized HTML page, is humongous; something on the order of going from 1ms to about 100ms.
